Removes excessive surgery time, tweaks medicine a lot

TobyThorne opened the pull request:
У получасового дроча в хирургию довольно сомнительная геймплейная ценность. Хирургу нужно полчаса тыкать кнопочки, а пациенту - залипать в чёрный квадрат. Не годится.

- Длительность операций значительно (полтора-два раза) сокращена.
- Decaying и Necrotic органы переименованы в Critical и Destroyed соответственно.
- Обычное восстановление органов теперь успешно лечит Critical-органы. Ранее их нужно было или пересаживать, или обрабатывать перидаксоном. О второй фиче никто толком не знал (на бегиннерсе - особенно), да и химики, варящие перидаксон - вещь редкая.
- Перидаксон при операции теперь и убирает статус Critical, и лечит орган. Нанесение намного быстрее, чем при использовании геля/бинтиков.
- Увеличен шанс успеха (5 -> 50) при использовании есворда в качестве лазерного скальпеля. - Добавлена возможность использовать лазерный резак для ампутации (да-да, без выстрелов).
- Убрана возможность НОРМАЛЬНО лечить внутренние органы соматическим гелем/бинтами.
- Бинтики-гели-etc лучше лечат ранки сами по себе. 
- Добавлены инструменты для починки органов - "organ fixer". Наполняются соматическим гелем, тратят по одной дозе геля на починку одного органа. Количество имеющегося геля отображается через Examine. У пустых фиксеров меняется спрайт.
Имеют три варианта:
1. Стандартный - находится в ящиках с инструментами. Вмещает 10 доз геля. Чинит один орган за раз. Им оборудованы борги-хирурги.
2. Улучшенный - изучается в РнД. Вмещает 20 доз геля. Чинит все органы в выбранной части тела. Им оборудованы улучшенные борги-хирурги.
Технологии: TECH_MATERIAL = 5, TECH_ENGINEERING = 4, TECH_BIO = 4
Ресурсы: MATERIAL_STEEL = 10000, MATERIAL_GLASS = 5000, MATERIAL_GOLD = 250, MATERIAL_SILVER = 750
3. Блюспейс - изучается в РнД. Не использует гель для лечения. Чинит все органы в выбранной части тела с повышенной скоростью.
Технологии: TECH_MATERIAL = 7, TECH_ENGINEERING = 6, TECH_BIO = 6, TECH_BLUESPACE = 5
Ресурсы: MATERIAL_STEEL = 12500, MATERIAL_GLASS = 7500, MATERIAL_GOLD = 1500, MATERIAL_SILVER = 2000, MATERIAL_DIAMOND = 750
- Добавлены баки с соматическим и ожоговым гелями. Содержат по 300 доз. Из них можно заправлять соответствующие баночки, а из соматического - ещё и органфиксер. Меняют спрайт, когда остаётся половина, либо пустые. 
- В карго добавлены ящики с соматическим/ожоговым гелем (по 5 банок, 10 очков) и с соответствующими баками (по 1 баку, 10 очков).
- Гели теперь меняют спрайт, если были использованы (три стейта: полный, использованный, пустой). 
- Гели можно наполнять друг из друга.
- Добавлено гетто-восстановление органов. Все варианты имеют пониженный шанс на успех, занимают вдвое больше времени на восстановление органа, чем фиксер и не до конца чинят Critical (Decaying) органы.
Соматический гель - 67%
Бинтик - 34%
Изолента - 20%

Попутно переделал спрайт cautery, ибо он был слишком смешной.
![cautery](https://user-images.githubusercontent.com/45202681/82846171-679bb080-9f09-11ea-94b0-f14bf779d0e3.png)
![organfixer](https://user-images.githubusercontent.com/45202681/82846175-68344700-9f09-11ea-9233-946d5cff90e8.png)
![geltanks](https://user-images.githubusercontent.com/45202681/82903681-1fb96f80-9f83-11ea-8739-699c824463f5.png)


- [x] Pull Request полностью завершен, мне не нужна помощь чтобы его закончить.
- [x] Я внимательно прочитал все свои изменения и багов в них не нашел.
- [x] Я запускал сервер со своими изменениями локально и все протестировал.
- [x] Я ознакомился c [Guide to Contribute](https://github.com/ChaoticOnyx/OnyxBay/blob/dev/docs/contributing.md).

Commit:
BayMed nerfed

- The duration of surgery steps has been significantly reduced;
- Decaying and Necrotic organs have been renamed into Critical and Destroyed respectively;
- Critical-state organs can now be successfully repaired into the normal state;
- Peridaxon surgery now both removes Critical status AND repairs organs, much faster than somatic gel;

Voldirs commented:
В целом норм, сокращение времени операций это реально то что нужно. Но переименовывание декэя в критикал, а некроза в дестроед это ерунда. Там была проблема в том что орган из дэкая в некроз переходил тупо по таймеру, крио никак не влияло, антибиотики тоже.

Epicus7 commented:
Касается #1784

Epicus7 commented on code/modules/surgery/generic.dm:
```diff
@@ -34,11 +34,11 @@
 	/obj/item/weapon/scalpel/laser2 = 85, \
 	/obj/item/weapon/scalpel/laser1 = 75, \
-	/obj/item/weapon/melee/energy/sword = 5
+	/obj/item/weapon/melee/energy/sword = 50
```
Про это в ОПе не написано

TobyThorne replied:
Добавил.

Epicus7 commented on code/modules/surgery/generic.dm:
```diff
@@ -297,11 +297,12 @@
 	/obj/item/weapon/material/hatchet = 75, \
-	/obj/item/weapon/material/twohanded/fireaxe = 85
+	/obj/item/weapon/material/twohanded/fireaxe = 85, \
+	/obj/item/weapon/gun/energy/plasmacutter = 90
```
Тоже в ОПе нету.

TobyThorne replied:
Добавил.

Epicus7 reviewed (changes requested):
ОП надо бы дополнить, некоторые геймплейные изменения не описаны.

TapKo4eB reviewed (dismissed):
ждём геймдизов

vekotov commented:
М.б. раз уж сделали перидаксон менее полезным, то увеличите ему рейты лечения при введении через шприц прямо в тело?

Moscowchief commented:
Не вижу смысла в переименовании состояния органов. Мне кажется Decaying и Necrotic звучат получше. 
> Обычное восстановление органов теперь успешно лечит Critical-органы. 
Перидаксон при операции теперь и убирает статус Critical, и лечит орган.

Я не понял, бинтик тоже убирает статус или что?

TobyThorne commented:
> Не вижу смысла в переименовании состояния органов. Мне кажется Decaying и Necrotic звучат получше.

Ну, тут вкусовое. Просто гниение органов за несколько минут/из-за урона выглядит как-то очень странно. Как будто каждый космонавтик - некролит. 
А "крит" - давно и часто используется в меде ССки.
*shrug

> Я не понял, бинтик тоже убирает статус или что?

Ага. Бинтик используется в операциях так же, как соматический гель (старый advanced brute pack), но со значительно сниженным шансом на успех. 
Но да, теперь, как написал в шапке, обычное восстановление органов лечит и Decaying (Critical) органы. Эти бесконечные замены на протезы и беготня за металлом - какой-то несмешной дроч. В случае отсутствия перидаксона (90% раундов), Decaying ничем не отличался от Necrotic и практически всем пациентам приходится пересаживать половину внутреннего мира.

Moscowchief commented:
А зачем тогда нужен перидаксон, если его функции выполняет бинтик?

TobyThorne commented:
> А зачем тогда нужен перидаксон, если его функции выполняет бинтик?

Хз. Его всё равно никто кроме Лосяра на моей памяти не использовал в операциях. Да и вообще его использовать смысла особого нет - это надо сначала бинтиками вылечить Decaying орган, а потом уже перидаксоном снимать статус. Оставил (и баффнул) операцию с ним в качестве быстрой альтернативы бинтику. 
В основном все его просто колют шприцем, когда органы чуть-чуть задамажены, либо во время операции дабы медленнее отваливались.

Rodial-Admiral commented:
Decaying - бесполезная ерунда, которую по факту проще пересадить, чем возиться с пипеткой и перидаксоном. Так что да, не нужен.

Rodial-Admiral commented:
Я бы ввел тогда восстановление некрозных органов ударными дозами перидаксона, чтобы это был честный сдвиг на одну ступень.

И да, чем теперь critical отличаются от просто продамаженных органов? Надо, чтобы какая-то механика у них была отличная от обычных.

Может, постепенное получение урона им сделать?

TobyThorne commented:
> И да, чем теперь critical отличаются от просто продамаженных органов? Надо, чтобы какая-то механика у них была отличная от обычных.
> Может, постепенное получение урона им сделать?

Они же, вроде, и так ловят постоянный дамаг, когда Decaying (Critical). Каждый орган - это как космонавтики из старых билдов. Упал в крит - начинает умирать.

Voldirs commented:
О, точняк, пускай тогда операция с периком оживляет полностью сдохшие органы. Главное чтоб мозг не восстанавливало

Rodial-Admiral commented:
Если критические уже сами потихоньку помирают - тогда остается момент с перидаксоном, который через пипетку мог бы оживлять некротические органы.

Тут еще вопрос - полностью оживлять, или выводить до верхней планки крита, чтобы надо было еще подорожником приложить.

@Moscowchief - ?

Moscowchief commented:
> выводить до верхней планки крита, чтобы надо было еще подорожником приложить.

@Rodial-Admiral This.

TapKo4eB commented:
Мне на самом деле всегда казалось стрёмным полное восстановление чего-то из мёртвого состояния, но пусть, ладно. По крайней мере если это воскрешение выводит до верхней планки крита, чтобы ещё приходилось домазывать гелем.

Commit:
Update organs_internal.dm

TobyThorne commented:
Добавил перидаксону воскрешение Destroyed органов до верхней планки крита, чтобы ещё приходилось домазывать гелем.

Commit:
Peridaxon buff

- 1 unit of peridaxon now heals 3 points of internal damage;

TobyThorne commented:
Короче, пусть висит, геймдизы умные, геймдизы разберутся.
- Баффнул перидаксон при внутривенном введении, теперь 1 единица перидаксона восстанавливает не 1, а 3 единицы урона по органам.

Commit:
hugesurgeryupdate

- Internal organs can no longer be fixed w/ somatic gel/bandages.
- Added three types of organ fixer tool:
- Standard organ fixer
- Advanced organ fixer
- Blusepace organ fixer
- Redrew cautery sprite

TobyThorne commented:
- Удалил возможность лечить внутренние органы соматическим гелем/бинтами.
- Добавлены инструменты для починки органов - "organ fixer". Наполняются соматическим гелем, тратят по одной дозе геля на починку одного органа.
Имеют три варианта:
- Стандартный - находится в ящиках с инструментами. Вмещает 10 доз геля. Чинит один орган за раз. Им оборудованы борги-хирурги.
- Улучшенный - изучается в РнД. Вмещает 20 доз геля. Чинит все органы в выбранной части тела. Им оборудованы улучшенные борги-хирурги.
Технологии: TECH_MATERIAL = 5, TECH_ENGINEERING = 4, TECH_BIO = 4
Ресурсы: MATERIAL_STEEL = 10000, MATERIAL_GLASS = 5000, MATERIAL_GOLD = 250, MATERIAL_SILVER = 750
- Блюспейс - изучается в РнД. Не использует гель для лечения. Чинит все органы в выбранной части тела с повышенной скоростью.
Технологии: TECH_MATERIAL = 6, TECH_ENGINEERING = 4, TECH_BIO = 6, TECH_BLUESPACE = 3
Ресурсы: MATERIAL_STEEL = 12500, MATERIAL_GLASS = 7500, MATERIAL_GOLD = 1500, MATERIAL_SILVER = 2000, MATERIAL_DIAMOND = 750

Попутно переделал спрайт cautery, ибо он был слишком смешной.
![cautery](https://user-images.githubusercontent.com/45202681/82846171-679bb080-9f09-11ea-94b0-f14bf779d0e3.png)
![organfixer](https://user-images.githubusercontent.com/45202681/82846175-68344700-9f09-11ea-9233-946d5cff90e8.png)

Commit:
quickfix

Commit:
quickfix2

TapKo4eB commented on code/game/objects/items/weapons/storage/uplink_kits.dm:
```diff
@@ -220,6 +220,7 @@
 		/obj/item/weapon/surgicaldrill,
 		/obj/item/weapon/bonegel,
 		/obj/item/weapon/FixOVein/clot,
+		/obj/item/weapon/organfixer/standard,
```
Так синди кит по задумке же должен быть спецом круче, почему бы в него не положить орган фиксер второго тира?

TobyThorne replied:
Granted.

TapKo4eB commented on code/modules/surgery/organs_internal.dm:
```diff
@@ -10,128 +10,232 @@
 
+/datum/surgery_step/internal/fix_organ_multiple/can_use(mob/living/user, mob/living/carbon/human/target, target_zone, obj/item/tool)
+	var/obj/item/weapon/organfixer/O = tool
+	if(!istype(O))
+		return FALSE
 	if (!hasorgans(target))
```
![изображение](https://user-images.githubusercontent.com/48761132/82866741-2cfa3e00-9f32-11ea-9489-ce825f34f718.png)

TobyThorne replied:
![image](https://user-images.githubusercontent.com/45202681/82880773-b163b580-9f60-11ea-8b58-1c031981d3ec.png)
Там ещё такое охуительное есть. Забыл вчера поправить.

TapKo4eB commented on code/game/objects/items/weapons/surgery_tools.dm:
```diff
@@ -262,3 +262,84 @@
+	name = "organ fixer"
+	desc = "A device used to fix internal organs."
+	icon = 'icons/obj/surgery.dmi'
+	item_state = "scientology" // TODO: Draw a proper handheld sprite; For now it looks fine ~Toby
```
![yey](https://user-images.githubusercontent.com/48761132/82868226-ee19b780-9f34-11ea-929a-ccfaa5637b7a.png)
better fix it because i can see it through my ass

TobyThorne replied:
Когда-нибудь, когда найдётся добрый спрайтер. Ну или могу вообще убрать item_state, как у большинства инструментов.

TapKo4eB commented on code/game/objects/items/weapons/surgery_tools.dm:
```diff
@@ -262,3 +262,84 @@
+		if(gel_amt >= gel_amt_max)
+			to_chat(user, SPAN("notice", "\The [src] is full."))
+			return
+		O.use(1)
```
Может быть сделать чтобы гель сразу под лимит заливался, а не по 10 раз закликивать аппарат?

FOxOTHuK replied:
Noice effect

TobyThorne replied:
Ну тако+е. Если есть последняя на станции банка геля и пустой фиксер, а на руках пациент с убитой печенью и несколькими кровотечениями - получится смешное.

TapKo4eB reviewed (commented):
![empty](https://user-images.githubusercontent.com/48761132/82869412-e0653180-9f36-11ea-8248-083cb8581fbe.png)
Это довольно странно что тауч вылезает при пустом аппарате. Это наверное тоже бай дезинг?

Voldirs commented on code/modules/surgery/organs_internal.dm:
```diff
@@ -10,112 +10,217 @@
 	allowed_tools = list(
-	/obj/item/stack/medical/advanced/bruise_pack= 100,	\
-	/obj/item/stack/medical/bruise_pack = 40,	\
-	/obj/item/weapon/tape_roll = 20
```
А зачем гетто варианты то убирать ну?

Voldirs reviewed (commented):
А ещё надо тогда в карго добавлять конкретно ящик с гелями

Voldirs commented:
А, вот ещё такой момент, есть подозрение что с дефолтными инструментами (а большую часть раундом они именно дефолтные ибо учёные заняты клепанием пушечек и вротексов) лечение отказа органов будет ещё больше чем сейчас потому что надо будет тыкать каааааааааждый орган чтобы вылечить человека. Короче последние изменения надо продумать получше.

TobyThorne committed:
Update uplink_kits.dm

TobyThorne committed:
Update organs_internal.dm

TobyThorne commented:
> А, вот ещё такой момент, есть подозрение что с дефолтными инструментами (а большую часть раундом они именно дефолтные ибо учёные заняты клепанием пушечек и вротексов) лечение отказа органов будет ещё больше чем сейчас потому что надо будет тыкать каааааааааждый орган чтобы вылечить человека. Короче последние изменения надо продумать получше.

Дык зато время применения вдвое сокращено, с 7 до 3.5 секунд. Если нужно починить два органа - будет то же, что и сейчас. Если три - дольше. Что, в общем-то, уже окупается ускорением остальных степов. Да и повод идти полоскать мозги учёным.

Commit:
somaticupdate

- Added somatic gel (5pcs) and burn gel (5pcs) crates to supply;
- Added somatic gel tanks and burn gel tanks;
- Added somatic gel tank and burn gel tank crates to supply;
- Somatic and burn gels can now be transfered from/to their respective containers;
- Improved bangade/oinment/somatic gel/burn gel healing;
- Added "half-used" icons for somatic and burn gels;

Commit:
Merge branch 'medfix' of https://github.com/TobyThorne/OnyxBay into medfix

koronoshka commented:
Можно такую к одиссею прихуячить? Просто так, ради прикола

Voldirs commented:
Ооооооокей, идея мне нравится, осталось только узнать что будет если кинуть в такой бак человека))) Блин, получается что-то вроде как система реагентов, но при этом к реагентам отношения не имеющая.

Commit:
ghetto organ repairs

- Returned the ability to fix internal organs with somatic gel, bandages and tape. All of them are now considered "ghetto" tools and have lowered chance of success, as well as longer time to use.

Commit:
сука говно

ебаное сохраняй изменения нормально

TobyThorne commented:
- Добавил гетто-восстановление органов. Все варианты имеют пониженный шанс на успех и занимают вдвое больше времени на восстановление органа, чем фиксер.
Соматический гель - 67%
Бинтик - 34%
Изолента - 20%

Voldirs commented:
Вот, это прям мегалайк

PlinHost commented on code/game/objects/items/weapons/surgery_tools.dm:
```diff
@@ -262,3 +262,84 @@
+
+/obj/item/weapon/organfixer/advanced/bluespace
+	name = "bluespace organ fixer"
+	desc = "A heavyly modified device, resembling QROF-26 produced by Vey-Mey. This prototype has some sort of bluespace-related device attached, and doesn't seem to have any gel injection ports."
```
>Vey-Mey

TobyThorne replied:
Это еврейский отдел Вей-Меда

Commit:
vey-mey abolition

- Vey-Mey has been disbanded, distributing the employees among the other branches of Vey-Med Company.

Epicus7 commented on code/game/objects/items/weapons/storage/firstaid.dm:
```diff
@@ -128,6 +128,7 @@
 		/obj/item/weapon/surgicaldrill,
 		/obj/item/weapon/bonegel,
 		/obj/item/weapon/FixOVein,
+		/obj/item/weapon/organfixer,
```
Не вижу в ОПе

TobyThorne replied:
А он есть
>Стандартный - находится в ящиках с инструментами

Epicus7 commented on code/game/objects/items/weapons/storage/firstaid.dm:
```diff
@@ -142,6 +143,7 @@
 		/obj/item/weapon/surgicaldrill,
 		/obj/item/weapon/bonegel,
 		/obj/item/weapon/FixOVein,
+		/obj/item/weapon/organfixer/standard,
```
Тоже нету в ОПе

TobyThorne replied:
А он есть
>Стандартный - находится в ящиках с инструментами

Epicus7 commented on code/game/objects/items/weapons/surgery_tools.dm:
```diff
@@ -262,3 +262,84 @@
+			to_chat(user, SPAN("notice", "\The [src] doesn't seem to be reloadable."))
+			return
+		var/obj/item/stack/medical/advanced/bruise_pack/O = I
+		if(O.stack_empty == 2 || !O.get_amount())
+			to_chat(user, SPAN("warning", "\The [O] is empty!"))
+			return
```
Все эти проверки уже реализованы в O.use(). Нужно просто проверять вывод этой функции.

Epicus7 commented on code/game/objects/items/weapons/surgery_tools.dm:
```diff
@@ -262,3 +262,84 @@
+		..()
+
+/obj/item/weapon/organfixer/examine(mob/user)
+	if(..(user, 0))
```
Зачем тут аргументы указывать? Просто ..() передаст все аргументы автоматически.

У нас уже были баги, когда в таких местах аргументы обрезались и происходит какой-то треш.

Epicus7 replied:
Причем это было как раз в examine :)

Даже ПР нашел: https://github.com/ChaoticOnyx/OnyxBay/pull/1497 . У тебя, скорее всего, тот же баг. Прикол в том, что в examine нынче передается миллион аргументов, а ты их тут обрезаешь. Если делать ..(), то весь этот миллион аргументов передастся сам и тебе даже не нужно самому их указывать в аргументах твоей функции.

Epicus7 commented on code/game/objects/structures/geltanks.dm:
```diff
@@ -0,0 +1,86 @@
+	atom_flags = ATOM_FLAG_CLIMBABLE
+
+/obj/structure/geltank/examine(mob/user)
+	. = ..()
```
Так, а почему у organfixer-а в экзамайне нет возвращаемого значения? Видимо там пропустил?

Epicus7 commented on code/game/objects/structures/geltanks.dm:
```diff
@@ -0,0 +1,86 @@
+	icon_state = "somatictank"
+
+/obj/structure/geltank/somatic/attackby(obj/item/W as obj, mob/user as mob)
+	if(istype(W, /obj/item/stack/medical/advanced/bruise_pack))
+		var/obj/item/stack/medical/advanced/bruise_pack/G = W
+		if(capacity <= 0)
+			to_chat(user, SPAN("warning", "\The [src] is empty!"))
+			return
+		if(G.amount >= G.max_amount)
+			to_chat(user, SPAN("notice", "\The [G] is already full."))
+			return
+		var/amt_to_transfer = min((G.max_amount - G.amount), capacity)
+		use(amt_to_transfer)
+		G.refill(amt_to_transfer)
+		to_chat(user, SPAN("notice", "You refill \the [src] with [amt_to_transfer] doses of somatic gel."))
```
Ты вот этот весь код копируешь три раза. А он ппц какой стремный. Надо все общее перенести в use методы geltank и stack.

Epicus7 commented on code/modules/mob/living/silicon/robot/robot_modules.dm:
```diff
@@ -281,6 +282,10 @@ var/global/list/robot_modules = list(
 		PS.reagents.add_reagent(/datum/reagent/acid/polyacid, 5 * amount)
 
+	var/obj/item/weapon/organfixer/standard/OF = locate(/obj/item/weapon/organfixer/standard) in src.modules
+	if(OF.gel_amt < OF.gel_amt_max)
+		OF.gel_amt++
```
А аргумент amount нам не нужен, да?

Алсо опять проверка эта вылезла, почему бы не запихать её в organfixer, сделав там какой-то метод для пополнения, если его еще нет.

Epicus7 commented on code/modules/mob/living/silicon/robot/robot_modules.dm:
```diff
@@ -349,6 +355,10 @@ var/global/list/robot_modules = list(
 	var/obj/item/weapon/reagent_containers/spray/cleaner/drone/SC = locate(/obj/item/weapon/reagent_containers/spray/cleaner/drone) in src.modules
 	SC.reagents.add_reagent(/datum/reagent/space_cleaner,10 * amount)
 
+	var/obj/item/weapon/organfixer/advanced/OF = locate(/obj/item/weapon/organfixer/advanced) in src.modules
+	if(OF.gel_amt < OF.gel_amt_max)
+		OF.gel_amt++
```
Тоже самое, игнорируем amount почему-то и проверку лучше запихать в organfixer

Epicus7 commented on code/modules/surgery/organs_internal.dm:
```diff
@@ -10,99 +10,319 @@
-
-	if (!hasorgans(target))
-		return 0
+	if(!hasorgans(target))
```
Все короче, на organs_internal.dm я сдаюсь.

Это пиздец, а не ПР, чувствую тонны копипасты здесь.

Дальше, вроде, ничего такого, что сломает прям билд нет - но насчет хирургии хуй знает. Можем залить конеш, но косяки наверняка будут и я надеюсь Тоби их своевременно пофиксит.

TobyThorne commented on code/game/objects/items/stacks/medical.dm:
```diff
@@ -102,7 +108,7 @@
 	origin_tech = list(TECH_BIO = 1)
 	slot_flags = SLOT_GLOVES
-	heal_brute = 1
+	heal_brute = 5
```
Поп равлено.

Commit:
refactor

- some OOP stuff

Commit:
Update organs_internal.dm

- removed extra checks; if(!affected) is being checked in /datum/surgery_step/internal/can_use()

Epicus7 commented on code/game/objects/items/stacks/medical.dm:
```diff
@@ -193,11 +193,25 @@
 	update_icon()
+	return 1
+
+/obj/item/stack/medical/advanced/proc/refill_from_same(obj/item/stack/medical/advanced/O, mob/user)
```
Не вижу внутри функции проверку на тип. Что мешает пополнить бинтики мазью?

Тут бы ассерт не помешал.

TobyThorne replied:
Это уже какой-то код ради кода, проверка снаружи идёт.

Epicus7 replied:
Это не код ради кода, у тебя прок называется "refill_from_same", а на самом деле он рефиллит даже из коробки с мешками для пылесоса, если у нас такая в билде, конечно, есть.

Твой прок обманывает кодера, который его использует. Не надо обманывать кодеров, так появляются баги.

Epicus7 replied:
Проверка аргументов функции - это ответственность самой функции.

TobyThorne replied:
переделал

Epicus7 commented on code/game/objects/items/stacks/medical.dm:
```diff
@@ -213,17 +227,8 @@
 
 /obj/item/stack/medical/advanced/bruise_pack/attackby(obj/item/W, mob/user)
 	if(istype(W, /obj/item/stack/medical/advanced/bruise_pack))
```
```suggestion
	var/obj/item/stack/medical/advanced/bruise_pack/O = W
	if(istype(O))
```

Тут можно не указывать тип два раза подряд. Изменять потом проще будет (или копировать, как мы любим).

TobyThorne replied:
переписал

Epicus7 commented on code/game/objects/items/stacks/medical.dm:
```diff
@@ -193,11 +193,25 @@
+	return 1
+
+/obj/item/stack/medical/advanced/proc/refill_from_same(obj/item/stack/medical/advanced/O, mob/user)
+	if(!O.amount)
```
```suggestion
	if(!O.get_amount())
```

И в остальных местах тоже. Иначе у тебя у киборгов все сломается.

TobyThorne replied:
поправил

Commit:
procs refactor

Epicus7 commented on code/game/objects/items/stacks/medical.dm:
```diff
@@ -202,7 +202,10 @@
 
-/obj/item/stack/medical/advanced/proc/refill_from_same(obj/item/stack/medical/advanced/O, mob/user)
+/obj/item/stack/medical/advanced/proc/refill_from_same(obj/item/I, mob/user)
+	if(!istype(src, I))
+		return
```
```suggestion
	ASSERT(istype(src, I))
```

Если твое условие выполнится - у нас очевидный баг, кто-то неправильно использует твою функцию. Но мы об этом никогда не узнаем, пока какой-то игрок не пожалуется, что у него от использования аптечки рука отвалилась или еще что-то страшное произошло.

ASSERT запишет рантайм и его будет сложно незаметить.

TobyThorne replied:
Если сюда воткнуть ассерт - то оно будет каждый раз срать рантаймами при попытке кликнуть на бёрнгель соматиком и наоборот.

Epicus7 replied:
Ну это можно было не переносить

Ладно, пофиг

TapKo4eB commented on code/game/objects/structures/geltanks.dm:
```diff
@@ -0,0 +1,81 @@
+	else if(capacity <= 150)
+		icon_state = "[initial(icon_state)]-used"
+
+/obj/structure/geltank/proc/fill_gel(obj/item/stack/medical/advanced/O, mob/user)
+	if(capacity <= 0)
+		to_chat(user, SPAN("warning", "\The [src] is empty!"))
+		return
+	var/amt_to_transfer = min((O.max_amount - O.amount), capacity)
+	if(O.refill(amt_to_transfer))
+		to_chat(user, SPAN("notice", "You refill \the [src] with [amt_to_transfer] doses of [gel_type] gel."))
+		use(amt_to_transfer)
+	else
+		to_chat(user, SPAN("notice", "\The [src] is already full."))
```
![durkablyat](https://user-images.githubusercontent.com/48761132/83419610-fc347000-a42d-11ea-8e57-c0fb167494c5.png)
`>прок заполнения гелевых стаканчиков из гигантскогого гель танка`
`>указываешь что ты заполнил гигантский гель танк вместо наоборот`

TobyThorne replied:
![cykablyat](https://user-images.githubusercontent.com/45202681/83421916-9bfff700-a44a-11ea-802d-034aa2b962e5.png)

TapKo4eB commented on code/game/objects/structures/geltanks.dm:
```diff
@@ -0,0 +1,81 @@
+		if(capacity <= 0)
+			to_chat(user, SPAN("warning", "\The [src] is empty!"))
+			return
+		if(G.gel_amt_max == -1)
+			to_chat(user, SPAN("notice", "\The [src] doesn't seem to be reloadable."))
+			return
+		var/amt_to_transfer = min((G.gel_amt_max - G.gel_amt), capacity)
+		if(G.refill(amt_to_transfer))
+			G.update_icon()
+			to_chat(user, SPAN("notice", "You refill \the [src] with [amt_to_transfer] doses of somatic gel."))
+			use(amt_to_transfer)
```
![durkaagain](https://user-images.githubusercontent.com/48761132/83420033-ab714700-a42e-11ea-820c-c1ef3aa35974.png)

Commit:
Update geltanks.dm

TobyThorne commented:
я шизик отвезите меня домой

TapKo4eB commented on code/modules/surgery/organs_internal.dm:
```diff
@@ -10,100 +10,312 @@
+		return FALSE
+	if(O.gel_amt == 0)
+		to_chat(user, SPAN("warning", "\The [O] is empty!"))
+		return FALSE
```
Во избежание протеканий проков и вывода двойных сообщений при операциях (а то и того хуже) в билде есть флаг SURGERY_FAILURE, который вешается на подобные ретёрны с сообщениями. Ради интереса посмотри код других операций. К слову проверь ещё раз все операции, может я где комментарий пропустил.

TapKo4eB commented on code/modules/surgery/organs_internal.dm:
```diff
@@ -10,100 +10,312 @@
+		return FALSE
+	if(target.op_stage.current_organ)
+		to_chat(user, SPAN("warning", "You can't do this right now."))
+		return FALSE
```
SURGERY_FAILURE

TapKo4eB commented on code/modules/surgery/organs_internal.dm:
```diff
@@ -10,100 +10,312 @@
+		return FALSE
+	if(!organ_to_fix.can_recover())
+		to_chat(user, SPAN("notice", "The [organ_to_fix.name] is destroyed and can't be saved."))
+		return FALSE
```
SURGERY_FAILURE

TapKo4eB commented on code/modules/surgery/organs_internal.dm:
```diff
@@ -10,100 +10,312 @@
+		return FALSE
+	if(!organ_to_fix.damage && !O.emagged)
+		to_chat(user, SPAN("notice", "The [organ_to_fix.name] is intact and doesn't require any healing."))
+		return FALSE
```
SURGERY_FAILURE

TapKo4eB commented on code/modules/surgery/organs_internal.dm:
```diff
@@ -10,100 +10,312 @@
 		return FALSE
+	if(O.gel_amt == 0)
+		to_chat(user, SPAN("warning", "\The [O] is empty!"))
+		return FALSE
```
SURGERY_FAILURE, к слову я могу сделать вывод что ты либо совсем невнимательный, либо делал этот код в 3 часа ночи, ибо даже ниже, в месте где ты проставлял испанцев есть этот самый флаг.

TapKo4eB commented on code/modules/surgery/organs_internal.dm:
```diff
@@ -10,100 +10,312 @@
+		return FALSE
+	if(target.op_stage.current_organ)
+		to_chat(user, SPAN("warning", "You can't do this right now."))
+		return FALSE
```
SURGERY_FAILURE

TapKo4eB commented on code/modules/surgery/organs_internal.dm:
```diff
@@ -10,100 +10,312 @@
+		var/obj/item/stack/medical/M = tool
+		if(M.amount < 1)
+			to_chat(user, SPAN("warning", "\The [M] is empty!"))
+			return FALSE
```
SURGERY_FAILURE

TapKo4eB commented on code/modules/surgery/organs_internal.dm:
```diff
@@ -10,100 +10,312 @@
+		return FALSE
+	if(!organ_to_fix.can_recover())
+		to_chat(user, SPAN("notice", "The [organ_to_fix.name] is destroyed and can't be saved."))
+		return FALSE
```
SURGERY_FAILURE

TapKo4eB commented on code/modules/surgery/organs_internal.dm:
```diff
@@ -10,100 +10,312 @@
+		return FALSE
+	if(!organ_to_fix.damage)
+		to_chat(user, SPAN("notice", "The [organ_to_fix.name] is intact and doesn't require any healing."))
+		return FALSE
```
SURGERY_FAILURE, а ещё ты ничего не тестил.

TapKo4eB commented on code/modules/surgery/organs_internal.dm:
```diff
@@ -188,27 +398,25 @@
-		return 0
+	if(target.op_stage.current_organ)
+		to_chat(user, SPAN("warning", "You can't do this right now."))
+		return FALSE
```
SURGERY_FAILURE, стыд.

TapKo4eB commented on code/modules/surgery/organs_internal.dm:
```diff
@@ -343,23 +554,24 @@
-	target.op_stage.current_organ = null
+	if(target.op_stage.current_organ)
+		to_chat(user, SPAN("warning", "You can't do this right now."))
+		return 0
```
SURGERY_FAILURE, было чуть выше.

TapKo4eB commented on code/modules/surgery/organs_internal.dm:
```diff
@@ -430,13 +642,12 @@
-	target.op_stage.current_organ = null
+	if(target.op_stage.current_organ)
+		to_chat(user, SPAN("warning", "You can't do this right now."))
+		return 0
```
SURGERY_FAILURE

TapKo4eB commented on code/modules/surgery/organs_internal.dm:
```diff
@@ -446,11 +657,8 @@
-		to_chat(user, "<span class='notice'>The [organ_to_fix.name] needs to be repaired before it is regenerated.</span>")
+	if(!organ_to_fix.can_recover() && istype(organ_to_fix, /obj/item/organ/internal/brain))
+		to_chat(user, SPAN("warning", "The [organ_to_fix.name] is destroyed and can't be saved."))
 		return 0
```
SURGERY_FAILURE, вот тут вообще смешно, ведь если не проставить этот флаг то прок будет проходить удачно, при этом нихрена не восстанавливая.

Commit:
Update organs_internal.dm

- Places SURGERY_FAILUREs where needed;
- Fixed the ability to call multiple internal organ pick-lists and proceed to operate NULL organs;

Commit:
Update organs_internal.dm

Rodial-Admiral commented:
Технологии: TECH_MATERIAL = 6, TECH_ENGINEERING = 4, TECH_BIO = 6, TECH_BLUESPACE = 3

Это очень маленькие уровни для блуспейс-тулзы, не требующей заправки.
Я бы предложил седьмой материал, шестой инжиниринг и пятый блуспейс. Био оставить на том же уровне.

TobyThorne committed:
Update dp-surgery.dm

- bluespace organ fixer now requires TECH_MATERIAL = 7, TECH_ENGINEERING = 4, TECH_BIO = 6, TECH_BLUESPACE = 5

TobyThorne commented:
> Технологии: TECH_MATERIAL = 6, TECH_ENGINEERING = 4, TECH_BIO = 6, TECH_BLUESPACE = 3
> 
> Это очень маленькие уровни для блуспейс-тулзы, не требующей заправки.
> Я бы предложил седьмой материал, шестой инжиниринг и пятый блуспейс. Био оставить на том же уровне.

Granted.

TobyThorne committed:
Update dp-surgery.dm

TECH_MATERIAL = 7, TECH_ENGINEERING = 6, TECH_BIO = 6, TECH_BLUESPACE = 5

TapKo4eB commented on code/game/objects/items/stacks/medical.dm:
```diff
@@ -172,30 +178,61 @@
 			affecting.disinfect()
 
+/obj/item/stack/medical/advanced/proc/refill(amt = 1)
+	if(get_amount() >= max_amount)
+		return 0
```
В чём смысл этой логики, и почему мы не можем заправлять орган фиксер если банка с гелью заполнена доверху?

Commit:
quickfix

- Fixed funny organ fixer refilling check;
- Advanced organ fixers assembled in RnD are now empty by default

TobyThorne commented:
- Поправил смешную проверочку
- Улучшенный фиксер из РнД теперь появляется пустым

TapKo4eB reviewed (approved):
*sigh
